psh/wget/wget.c

```diff
@@ -0,0 +1,485 @@
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <ctype.h>
+#include <errno.h>
+#include <netdb.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <unistd.h>
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <sys/minmax.h>
+#include <sys/time.h>
+
+#include "../psh.h"
+
+
+static struct {
+	char buf[4096];
+	char *readptr;
+	int len;
+	FILE *outfile;
+	const char *path;
+	char *host;
+	const char *filename;
+	int debug;
+	int socket;
+} common;
+
+
+void psh_wgetinfo(void)
+{
+	printf("downloads a file using http");
+}
+
+
+static void wget_help(void)
+{
+	printf("Usage: wget [options] ... URL\n");
+	printf("Options\n");
+	printf("  -h:  prints help\n");
+	printf("  -O:  output file\n");
+}
+
+
+static int wget_parseUrl(const char *url)
+{
+	const char httpProto[] = "http://";
+	const char *tmp, *bare;
+
+	if (strstr(url, "://") == NULL) {
+		/* No protocol specified, assume HTTP */
+		bare = url;
+	}
+	else if (strncmp(url, httpProto, strlen(httpProto)) == 0) {
+		bare = url + strlen(httpProto);
+	}
+	else {
+		printf("wget: Unsupported protocol!\n");
+		return -1;
+	}
+
+	tmp = strrchr(bare, '/');
+	if (tmp == NULL || tmp[1] == '\0') {
+		fprintf(stderr, "wget: url missing filename!\n");
+		return -1;
+	}
+
+	common.filename = tmp + 1;
+
+	/* TODO: Decode URL */
+	tmp = strchr(bare, '/');
+	common.path = tmp + 1;
+	common.host = strndup(bare, tmp - bare);
+	if (common.host == NULL) {
+		fprintf(stderr, "wget: Out of memory!\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+
+static uint32_t *wget_aiToAddr(struct addrinfo *ai)
+{
+	return &(((struct sockaddr_in *)ai->ai_addr)->sin_addr.s_addr);
+}
+
+
+static int wget_connect(const char *host)
+{
+	struct addrinfo *res;
+	struct addrinfo hints = { 0 };
+	char hostaddr[INET_ADDRSTRLEN];
+	int fd = -1;
+
+	hints.ai_family = AF_INET;
+	hints.ai_socktype = SOCK_STREAM;
+	hints.ai_protocol = IPPROTO_TCP;
+
+	printf("Resolving %s... ", host);
+	fflush(stdout);
+	if (getaddrinfo(host, "80", &hints, &res) != 0) {
+		printf("Failed\n");
+		return -1;
+	}
+
+	if (inet_ntop(res->ai_family, wget_aiToAddr(res), hostaddr, sizeof(hostaddr)) == NULL) {
+		fprintf(stderr, "nslookup: Error while converting address!\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("%s\n", hostaddr);
+
+	printf("Connecting to %s|%s|:80... ", host, hostaddr);
+	fflush(stdout);
+
+	fd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+	if (fd < 0) {
+		printf("Failed\n");
+		freeaddrinfo(res);
+		return -1;
+	}
+
+	if (connect(fd, res->ai_addr, res->ai_addrlen) < 0) {
+		close(fd);
+		freeaddrinfo(res);
+		return -1;
+	}
+	printf("Connected\n");
+
+	freeaddrinfo(res);
+
+	return fd;
+}
+
+
+static int wget_request(void)
+{
+	char buf[256];
+	int bytes = 0, i = 0, ret;
+
+	bytes = snprintf(buf, sizeof(buf), "GET /%s HTTP/1.1\r\nHost: %s\r\nUser-Agent: Wget\r\n\r\n",
+		common.path, common.host);
+	if (bytes > sizeof(buf)) {
```

Comment:
```suggestion
	if ((bytes < 0) || (bytes >= sizeof(buf)) {
```
`bytes < 0` - error
`bytes >= sizeof(buf)` - not enough space